02-microarray/pathway-analysis_microarray_04_gsva.Rmd

```diff
@@ -0,0 +1,631 @@
+The pathway-level score calculated is a way of asking how genes _within_ a gene set vary as compared to genes that are _outside_ of that gene set ([Malhotra. 2018](https://towardsdatascience.com/decoding-gene-set-variation-analysis-8193a0cfda3)). 
+(This is sometimes called a competitive test in gene set enrichment literature.)
+The intuition here is that we will get pathway-level scores for each sample that indicate if genes in a pathway vary concordantly in one direction (over-expressed or under-expressed relative to the overall population) ([Hänzelmann2013](https://doi.org/10.1186/1471-2105-14-7)).
+
+The output is a gene set by sample matrix of GSVA scores.
+
+### Perform GSVA
+
+**REVIEW**
+
+```{r}
+gsva_results <- gsva(matrix,
+  hallmarks_list,
+  method = "gsva",
+  # Appropriate for our transformed data on log2-like scale
+  kcdf = "Gaussian",
+  # Minimum gene set size
+  min.sz = 5,
+  # Maximum gene set size
+  max.sz = 50,
+  # Compute Gaussian-distributed scores
+  mx.diff = TRUE,
+  # Print out information about each calculation step
+  verbose = TRUE
+)
+```
+
+**DRAFT**
+
+Note that the `gsva()` function documentation says we can use `kcdf = "Gaussian"` if we had log-CPMs, log-RPKMs or log-TPMs, but we would use `kcdf = "Poisson"` on integer counts.
+See `?gsva` for more options.
+
+```{r}
+# Let's explore what the output of gsva() looks like
+head(gsva_results)
+```
+
+## Visualizing Results
+
+### Violin plot
+
+Now let's make a violin plot to look at the distribution of scores.
+First we need to get this data in an appropriate format for `ggplot2`.
+
+**REVIEW**
+
+```{r}
+# The results are a matrix
+long_df <- gsva_results %>%
+  data.frame() %>%
+  # Gene set names are rownames
+  tibble::rownames_to_column("gene_set") %>%
+  # Get into long format
+  tidyr::pivot_longer(
+    cols = -gene_set,
+    names_to = "Kids_First_Biospecimen_ID",
+    values_to = "gsva_score"
+  ) %>%
+  dplyr::arrange(dplyr::desc(gsva_score))
+```
+
+Let's make a violin plot so we can look at the distribution of scores for each pathway.
+
+**REVIEW**
+
+```{r}
+# Violin plot comparing GSVA scores of different random gene set sizes
+violin_plot <- long_df %>%
+  ggplot2::ggplot(ggplot2::aes(
+    x = gene_set,
+    y = gsva_score
+  )) +
+  # Make a violin plot that's a pretty blue!
+  ggplot2::geom_violin(fill = "#99CCFF", alpha = 0.5) +
+  # Add a point with the mean value
+  ggplot2::stat_summary(
+    geom = "point",
+    fun = "mean",
+    # Change the aesthetics of the points
+    size = 3,
+    color = "#0066CC",
+    shape = 18
+  ) +
+  # Flip the axes
+  ggplot2::coord_flip() +
+  ggplot2::labs(
+    title = "Gene Set GSVA Scores",
+    x = "gene set",
+    y = "GSVA score"
+  ) +
+  ggplot2::theme_bw()
+
+# Display plot
+violin_plot
+```
+
+**DRAFT**
+
+What do you notice about these distributions?
+How might you use this information to inform your interpretation of GSVA scores?
```

Comment:
I know this text is marked as a draft, but I want to make a larger point about the purpose of this plot. 

This looks like it's taken from the training material. In the training material, we generate a distribution of scores using random gene sets of different sizes to specifically illustrate the point about how the number of genes in a pathway/gene set influences the scores. I'm not sure what this violin plot is intended to illustrate if we're not going to include information about the gene set sizes. I don't know that we want to include that same experiment from training here (in fact I think we _don't_ want to include it). Instead, a more common use case for a violin plot is to look at the differences between groups. If you want to do that, you probably need to pick a different dataset with more metadata based on your comment when you filed the PR. Another idea would be to include the gene set size in the labels (e.g., `HALLMARK_ANGIOGENESIS (<number of genes>)`).

Reply:
I did think that the gene set size experiment from training would not be well suited here but was not sure of the best alternative here, that said, I plan to hopefully find another dataset with more metadata. However, if that ends up being a bit more time-consuming than we would want, I'll go ahead and included the gene set size in the labels.